src/tir/transforms/storage_flatten.cc

```diff
@@ -137,8 +152,15 @@ class BufferShapeLegalize : public StmtExprMutator {
           << "Inconsistent dimensions for buffer " << op->buffer->name;
 
       Array<PrimExpr> new_indices;
-      for (size_t i = 0; i < entry.realized_begins.size(); i++) {
-        new_indices.push_back(op->indices[i] - entry.realized_begins[i]);
+
+      // Pad leading indices with zero, matching the "fuzzy_match"
+      // behavior from ArgBinder::BindBuffer.
+      size_t diff = entry.realized_begins.size() - op->indices.size();
+      for (size_t i = 0; i < diff; i++) {
+        new_indices.push_back(0);
```

Comment:
Assuming this is for matching cases like [1, 1, n, m], do we need to check that the leading axes are indeed extent=1?

Reply:
Let me ask a more general question, is it possible to expand or squeeze all shapes and then do exact matching as before? Noticing a fair amount of special casing in this commit for handling the extra unit dimensions.

Reply:
For the specific case, there is a check as part of the call to `ArgBinder::BindBuffer`.  I'll add a comment to indicate that.

For the general question, that was the intent in `BufferShapeLegalize`, so that afterwards the buffers all have a single well-defined shape.  It looks like I had missed one case where `BufferBindUnwrapper` changed the number of dimensions when binding to an external buffer, but I've added another commit to this PR to pull that behavior into `BufferShapeLegalize` instead.